File: sorting/go/sorting.go
package main

import (
	"fmt"
	"math/rand"
	"time"
)

func main() {
	test()
}

func selectionSort(arr []int) []int {
	for i := 0; i < len(arr); i++ {
		min := i
		for j := i + 1; j < len(arr); j++ {
			if arr[j] < arr[min] {
				min = j
			}
		}
		arr[i], arr[min] = arr[min], arr[i]
	}
	return arr
}

func bubbleSort(arr []int) []int {
	for i := 0; i < len(arr); i++ {
		for j := len(arr) - 1; j > i; j-- {
			if arr[j] < arr[j-1] {
				arr[j], arr[j-1] = arr[j-1], arr[j]
			}
		}

	}
	return arr
}

func insertionSort(arr []int) []int {
	for i := 1; i < len(arr); i++ {
		current := arr[i]
		j := i - 1
		for j >= 0 && arr[j] > current {
			arr[j+1] = arr[j]
			j--

		}
		arr[j+1] = current
	}
	return arr
}
func merge(left_arr []int, right_arr []int) []int {
	left_index := 0
	right_index := 0
	result := make([]int, len(left_arr)+len(right_arr))
	for left_index < len(left_arr) && right_index < len(right_arr) {
		if left_arr[left_index] < right_arr[right_index] {
			result[left_index+right_index] = left_arr[left_index]
			left_index++
		} else {
			result[left_index+right_index] = right_arr[right_index]
			right_index++
		}

	}
	for left_index < len(left_arr) {
		result[left_index+right_index] = left_arr[left_index]
		left_index++
	}
	for right_index < len(right_arr) {
		result[left_index+right_index] = right_arr[right_index]
		right_index++
	}

	return result
}

// using sentinel
func merge2(left_arr []int, right_arr []int) []int {
	left_arr = append(left_arr, 999999999)
	right_arr = append(right_arr, 999999999)
	left_index := 0
	right_index := 0
	result := make([]int, len(left_arr)+len(right_arr))
	for i := 0; i < len(result); i++ {
		if left_arr[left_index] < right_arr[right_index] {
			result[i] = left_arr[left_index]
			left_index++
		} else {
			result[i] = right_arr[right_index]
			right_index++
		}

	}
	return result
}

func mergeSort(arr []int) []int {
	if len(arr) <= 1 {
		return arr
	}
	mid := len(arr) / 2
	left_arr := mergeSort(arr[:mid])
	right_arr := mergeSort(arr[mid:])
	return merge(left_arr, right_arr)
}
func isSorted(arr []int) bool {
	for i := 1; i < len(arr); i++ {
		if arr[i] < arr[i-1] {
			return false
		}
	}
	return true
}

func test() {
	rand.Seed(time.Now().UnixNano())
	start := time.Now()
	// Testing with random numbers
	for i := 0; i < 10; i++ {
		arr := make([]int, 100)
		for j := 0; j < 100; j++ {
			arr[j] = rand.Intn(1000)
		}

		if !isSorted(selectionSort(arr)) {
			fmt.Println("Selection Sort failed")
		}
		if !isSorted(bubbleSort(arr)) {
			fmt.Println("Bubble Sort failed")
		}

		if !isSorted(insertionSort(arr)) {
			fmt.Println("Insertion Sort failed")
		}

		if !isSorted(mergeSort(arr)) {
			fmt.Println("Merge Sort failed")
		}

	}
	// print time
	fmt.Println("Time elapsed: ", time.Since(start))

}

File: sorting/Java/sorting.java
public class SortingAlgorithms {
    public static void main(String[] args) {
        System.out.println("bubble sort");
        int[] arr = { 5, 4, 3, 2, 1 };
        bubbleSort(arr);
        for (int i = 0; i < arr.length; i++) {
            System.out.print(arr[i]);
        }
        System.out.println("insertion sort");
        // insertionSort
        int[] arr2 = { 5, 4, 3, 2, 1 };
        insertionSort(arr2);
        for (int i = 0; i < arr2.length; i++) {
            System.out.print(arr2[i]);
        }
        System.out.println("merge sort");

        // mergeSort
        int[] arr3 = { 5, 4, 3, 2, 1 };
        arr3 = mergeSort(arr3);
        for (int i = 0; i < arr3.length; i++) {
            System.out.print(arr3[i]);
        }
    }

    public static int[] bubbleSort(int[] arr) {
        for (int i = 0; i < arr.length; i++) {
            for (int j = arr.length - 1; j > i; j--) {
                if (arr[j] < arr[j - 1]) {
                    int temp = arr[j];
                    arr[j] = arr[j - 1];
                    arr[j - 1] = temp;
                }
            }
        }
        return arr;
    }

    // sort from end bubbleSort
    public static int[] bubbleSort2(int[] arr) {
        for (int i = arr.length - 1; i >= 0; i--) {
            for (int j = 0; j < i; j++) {
                if (arr[j] > arr[j + 1]) {
                    int temp = arr[j + 1];
                    arr[j + 1] = arr[j];
                    arr[j] = temp;
                }
            }
        }
        return arr;
    }

    // selection sort
    public static int[] selectionSort(int[] arr) {
        for (int i = 0; i < arr.length; i++) {
            int min = i;
            for (int j = i + 1; j < arr.length; j++) {
                if (arr[j] < arr[min]) {
                    min = j;
                }
            }
            int temp = arr[min];
            arr[min] = arr[i];
            arr[i] = temp;
        }
        return arr;
    }

    public static int[] insertionSort(int[] arr) {
        for (int i = 1; i < arr.length; i++) {
            int j = i;
            int current = arr[i];
            while (j > 0 && arr[j - 1] > current) {
                arr[j] = arr[j - 1];
                j--;
            }
            arr[j] = current;
        }
        return arr;
    }

    public static int[] merge(int[] left_arr, int[] right_array) {
        int left_index = 0;
        int right_index = 0;
        int[] result = new int[left_arr.length + right_array.length];
        while (left_index < left_arr.length && right_index < right_array.length) {
            if (left_arr[left_index] < right_array[right_index]) {
                result[left_index + right_index] = left_arr[left_index];
                left_index++;
            } else {
                result[left_index + right_index] = right_array[right_index];
                right_index++;
            }
        }
        while (left_index < left_arr.length) {
            result[left_index + right_index] = left_arr[left_index];
            left_index++;
        }
        while (right_index < right_array.length) {
            result[left_index + right_index] = right_array[right_index];
            right_index++;
        }
        return result;
    }

    public static int[] mergeSort(int[] arr) {
        if (arr.length <= 1) {
            return arr;
        }
        int mid = arr.length / 2;
        int[] left_arr = new int[mid];
        int[] right_arr = new int[arr.length - mid];
        for (int i = 0; i < mid; i++) {
            left_arr[i] = arr[i];
        }
        for (int i = mid; i < arr.length; i++) {
            right_arr[i - mid] = arr[i];
        }
        left_arr = mergeSort(left_arr);
        right_arr = mergeSort(right_arr);

        return merge(left_arr, right_arr);

    }

}